library/core/src/pin.rs

```diff
@@ -1,188 +1,588 @@
+//! stay *un-moved* and valid during the address-sensitive portion of its lifetime because some
+//! interface is relying on those invariants to be true in order for its implementation to be sound.
+//!
+//! As a motivating example of a type which may become address-sensitive, consider a type which
+//! contains a pointer to another piece of its own data, *i.e.* a "self-referential" type. In order
+//! for such a type to be implemented soundly, the pointer which points into `self`'s data must be
+//! proven valid whenever it is accessed. But if that value is *moved*, the pointer will still
+//! point to the old address where the value was located and not into the new location of `self`,
+//! thus becoming invalid. A key example of such self-referrential types are the state machines
+//! generated by the compiler to implement [`Future`] for `async fn`s.
+//!
+//! Such types that have an *address-sensitive* state usually follow a lifecycle
+//! that looks something like so:
+//!
+//! 1. A value is created which can be freely moved around.
+//!     * e.g. calling an async function which returns a state machine implementing [`Future`]
+//! 2. An operation causes the value to depend on its own address not changing
+//!     * e.g. calling [`poll`] for the first time on the produced [`Future`]
+//! 3. Further pieces of the safe interface of the type use internal [`unsafe`] operations which
+//! assume that the address of the value is stable
+//!     * e.g. subsequent calls to [`poll`]
+//! 4. Before the value is invalidated (e.g. deallocated), it is *dropped*, giving it a chance to
+//! notify anything with pointers to itself that those pointers will be invalidated
+//!     * e.g. [`drop`]ping the [`Future`]
+//!
+//! There are two possible ways to ensure the invariants required for 2. and 3. above (which
+//! apply to any address-sensitive type, not just self-referrential types) do not get broken.
+//!
+//! 1. Have the value detect when it is moved and update all the pointers that point to itself
+//! 2. Guarantee that the address of the value does not change (and that memory is not re-used
+//! for anything else) during the time that the pointers to it are expected to be valid to
+//! dereference
+//!
+//! Since, as we discussed, Rust can move values without notifying them that they have moved, the
+//! first option is ruled out.
+//!
+//! In order to implement the second option, we must in some way enforce its key invariant,
+//! *i.e.* prevent the value from being *moved* or otherwise invalidated. (You may notice this
+//! sounds an awful lot like the definition of *pinning* a value). There are two ways by
+//! which we might enforce this validity invariant in Rust:
+//!
+//! 1. Offer a wholly `unsafe` API to interact with the object, thus requiring every caller to
+//! uphold the invariant themselves; or,
+//! 2. Leverage the type system to encode and enforce this invariant by presenting a restricted
+//! API surface to interact with the object
```

Comment:
I've added a section describing this, I think it is valuable although does make things even longer. :D

Reply:
I think you can make it shorter by omitting the "Each individual developer must themselves think through all the guarantees" part — it's _also_ true that developers seeking to rely on `Pin` need to think carefully, so it's not a unique disadvantage. Arguably Pin is better because it admits things like `pin!` and `pin_project`, but I think this section would work fine _just_ discussing the extra allocations.

Reply:
Indeed, however what I wanted to express is that when a developer uses `Pin` they do not need to think about whether the safe interface of `Pin` might allow unsoundness -- they only need to think about whether they're upholding their side of the contract with the `Pin` api. When making your own type, you not only have to think about whether you're actually doing a move, but also whether the safe api you expose is sound based on the unique invariants you choose to uphold or not. Perhaps I should try to reword that section again to make that more clear....

Reply:
Okay, I've trimmed this down as you suggested as I think it's not super important to elaborate more.